### ISSUE TITLE
Nominated node in cache didn't match Pod status

### ISSUE THREAD

**Issue by alculquicondor** [kubernetes/kubernetes#109236]
### What happened?

The scheduler cache entered a corrupted state where a Pod, according to the cache dump, was nominated to node A, while the status said it was nominated to node B.

This lead to unschedulable low priority pods due to the phantom pod in the nominator cache.

### What did you expect to happen?

The cache to match the Pod status

### How can we reproduce it (as minimally and precisely as possible)?

Setup: two nodes with all the resources used. A high priority pod comes in that is able to preempt at least one pod.

This might cause these sequence of events

1. Pod X doesn't fit, it gets nominated to `node-y` and added back to scheduling queue.
2. We start a second scheduling attempt for Pod X.
3. We get a Pod update for the `.status.nominatedPodName`. This puts the pod back into the scheduling queue.
4. The second scheduling attempt finishes and schedules the pod successfully in `node-y`
5. We start a third scheduling attempt for Pod X (due to 3)
6. We get a Pod update for the `.spec.nodeName`. This should have cleared the nominated node in the cache, but there is nothing currently there.
7. In the third scheduling attempt, the pod doesn't fit, so the scheduler nominates it to `node-z`. Nothing clears this nomination afterwards

Usually step 2 happens after 3, so everything works as expected. But in a busy cluster, 3 might be delayed.

### Kubernetes version

`1.21.9`

### Cloud provider

<details>
GKE
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by alculquicondor** [comment]
/sig scheduling
cc @Huang-Wei

**Comment by alculquicondor** [comment]
This Pod actually got scheduled:

```
2022-03-30 10:05:32.162 EDT
""Unable to schedule pod; no fit; waiting" pod="x" err="0/7 nodes are available: 1 node(s) had taint {ToBeDeletedByClusterAutoscaler: 1648649131}, that the pod didn't tolerate, 6 Insufficient cpu.""
Error
2022-03-30 10:05:32.162 EDT
""Error occurred" err="Pod x is already present in the active queue""
Info
2022-03-30 10:05:32.197 EDT
""Successfully bound pod to node" pod="x" node="y" evaluatedNodes=7 feasibleNodes=1"
Info
2022-03-30 10:05:32.226 EDT
""Unable to schedule pod; no fit; waiting" pod="x" err="0/7 nodes are available: 1 node(s) had taint {ToBeDeletedByClusterAutoscaler: 1648649131}, that the pod didn't tolerate, 6 Insufficient cpu.""
Info
2022-03-30 10:05:32.226 EDT
""Pod has been assigned to node. Abort adding it back to queue." pod="x" node="node-y""
```

The cache dump was obtained much later at `0331 14:21:53.634291`:

```
Node name: node-z
Scheduled Pods(number: 10):
...
Nominated Pods(number: 1):
name: x, namespace: default, uid: 8aa162b9-ef7f-4b50-a64c-4abd8e8aeb75, phase: Pending, nominated node: node-y
I0331 14:21:53.634530      12 dumper.go:49] 
```

I think the problem is that we got an update for the Pod (probably the nominated node itself), scheduled it again and assigned a second nominated node for it. The second binding failed, but we didn't clear the cache.

**Comment by Huang-Wei** [comment]
Sounds like #106816 is related?

**Comment by alculquicondor** [comment]
It could be the same, or it could be another bug. Note that this is a second scheduling attempt.

**Comment by alculquicondor** [comment]
My theory of events:

1. Pod X doesn't fit, it gets nominated to `node-y` and added back to scheduling queue.
2. We start a second scheduling attempt for Pod X.
3. We get a Pod update for the `.status.nominatedPodName`. This puts the pod back into the scheduling queue.
4. The second scheduling attempt finishes and schedules the pod successfully in `node-y`
5. We start a third scheduling attempt for Pod X (due to 3)
6. We get a Pod update for the `.spec.nodeName`. This should have cleared the nominated node in the cache, but there is nothing currently there.
7. In the third scheduling attempt, the pod doesn't fit, so the scheduler nominates it to `node-z`. Not sure why this update fails, but nevertheless, we have a corrupted cache at this point.

**Comment by Huang-Wei** [comment]
@alculquicondor thanks for raising up this and the fix. Could you update the "How can we reproduce it (as minimally and precisely as possible)?" section a bit?

Per our offline conversation, the issue will leave a "phantom" pod holding a number of resources in the internal nominator cache. And due to that unnecessary resource reservation, low-priority pods may not get scheduled while they should have been.

> 2. We start a second scheduling attempt for Pod X.

This might be the key trigger of this issue. In practice, it can be caused by a (way-too-quick) movePodBackToActiveQ call (due to relevant/irrelevant events, or backoff's completion) after step 1 and before step 3.

**Comment by alculquicondor** [comment]
Updated the description. I will update the PRs on Monday.
